text/0000-safe-mem-forget.md

```diff
@@ -0,0 +1,124 @@
+- Feature Name: N/A
+- Start Date: 2015-04-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Alter the signature of the `std::mem::forget` function to remove `unsafe`.
+Explicitly state that it is not considered unsafe behavior to not run
+destructors.
+
+# Motivation
+
+It was [recently discovered][scoped-bug] by @arielb1 that the `thread::scoped`
+API was unsound. To recap, this API previously allowed spawning a child thread
+sharing the parent's stack, returning an RAII guard which `join`'d the child
+thread when it fell out of scope. The join-on-drop behavior here is critical to
+the safety of the API to ensure that the parent does not pop the stack frames
+the child is referencing. Put another way, the safety of `thread::scoped` relied
+on the fact that the `Drop` implementation for `JoinGuard` was *always* run.
+
+[scoped-bug]: https://github.com/rust-lang/rust/issues/24292
+
+The [underlying issue][forget-bug] for this safety hole was that it is possible
+to write a version of `mem::forget` without using `unsafe` code (which drops a
+value without running its destructor). This is done by creating a cycle of `Rc`
+pointers, leaking the actual contents. It [has been pointed out][dtor-comment]
+that `Rc` is not the only vector of leaking contents today as there are
+[known][dtor-bug1] [bugs][dtor-bug2] where `panic!` may fail to run
+destructors. Furthermore, it has [also been pointed out][drain-bug] that not
+running destructors can affect the safety of APIs like `Vec::drain_range` in
+addition to `thread::scoped`.
+
+[forget-bug]: https://github.com/rust-lang/rust/issues/24456
+[dtor-comment]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93505374
+[dtor-bug1]: https://github.com/rust-lang/rust/issues/14875
+[dtor-bug2]: https://github.com/rust-lang/rust/issues/16135
+[drain-bug]: https://github.com/rust-lang/rust/issues/24292#issuecomment-93513451
+
+It has never been a guarantee of Rust that destructors for a type will run, and
+this aspect was overlooked with the `thread::scoped` API which requires that its
+destructor be run! Reconciling these two desires has lead to a good deal of
+discussion of possible mitigation strategies for various aspects of this
+problem. This strategy proposed in this RFC aims to fit uninvasively into the
+standard library to avoid large overhauls or destabilizations of APIs.
+
+# Detailed design
+
+Primarily, the `unsafe` annotation on the `mem::forget` function will be
+removed, allowing it to be called from safe Rust. This transition will be made
+possible by stating that destructors **may not run** in all circumstances (from
+both the language and library level). The standard library and the primitives it
+provides will always attempt to run destructors, but will not provide a
+guarantee that destructors will be run.
+
```

Comment:
I think we can guarantee much more. Let's use ownership as the core concept.

"The owner of a value has the final influence on whether the destructors run or not."

- You must own the value to forget it
- Your own owned values that you keep in a scope as guards will be guaranteed to drop as usual when they leave the scope
- Shared ownership leaves you at mercy of all the owners together.
- If you want to be sure the destructor runs, don't pass on ownership (like thread::scoped would return a value, relinquishing ownership).